stores/metadata.go

```diff
@@ -1728,6 +1765,74 @@ func (s *SQLStore) DeleteHostSector(ctx context.Context, hk types.PublicKey, roo
 	return deletedSectors, err
 }
 
+func (s *SQLStore) dirID(tx *gorm.DB, dirPath string) (uint, error) {
+	if !strings.HasPrefix(dirPath, "/") {
+		return 0, fmt.Errorf("path must start with /")
+	} else if !strings.HasSuffix(dirPath, "/") {
+		return 0, fmt.Errorf("path must end with /")
+	}
+
+	if dirPath == "/" {
+		return 1, nil // root dir returned
+	}
+
+	var dir dbDirectory
+	if err := tx.Where("name", dirPath).
+		Select("id").
+		Take(&dir).
+		Error; err != nil {
+		return 0, fmt.Errorf("failed to fetch directory: %w", err)
+	}
+	return dir.ID, nil
+}
+
+func makeDirsForPath(tx *gorm.DB, path string) (uint, error) {
+	// Create root dir.
+	dirID := uint(rootDirID)
+	if err := tx.Model(&dbDirectory{}).
+		Clauses(clause.OnConflict{
+			DoNothing: true,
+		}).Create(map[string]any{
+		"id":   dirID,
+		"name": "/",
+	}).Error; err != nil {
+		return 0, fmt.Errorf("failed to create root directory: %w", err)
+	}
+
+	// Create remaining directories.
+	path = strings.TrimSuffix(path, "/")
+	if path == "/" {
+		return dirID, nil
+	}
+	for i := 0; i < utf8.RuneCountInString(path); i++ {
+		if path[i] != '/' {
+			continue
+		}
+		dir := path[:i+1]
+		if dir == "/" {
+			continue
+		}
+		if err := tx.Clauses(clause.OnConflict{
+			DoNothing: true,
+		}).
+			Create(&dbDirectory{
+				Name:       dir,
+				DBParentID: dirID,
+			}).Error; err != nil {
+			return 0, fmt.Errorf("failed to create directory %v: %w", dir, err)
+		}
+		var childID uint
+		if err := tx.Raw("SELECT id FROM directories WHERE name = ?", dir).
```

Comment:
I think this is worth optimising to have faster migration but also to have a faster `UpdateObject`. Can't you save this query by doing `db := &dbDirectory{...} tx.Create(&db)` and then just pluck the id from `db`.
Also if you cache these you can speed this up I think because now in a regular folder structure type database we are recreating the same dirs over and over again and `DoNothing` solves it but we might as well just cache on whatever unique key it has and prevent doing the insert and fetching the `childID`

Reply:
`Create` doesn't always work with `ON CONFLICT` which is also the reason why we fetch sector ids after inserting sectors. Overall I wouldn't worry about this method cause it doesn't even show up since it's so fast. Pretty sure the ids that were just inserted are somehow cached anyway because even for large objects with many sectors I have never seen a slow txn.